src/hash.hpp

```diff
@@ -13,115 +15,138 @@
 #include "common.h"
 #include "cyclic.hpp"
 
-#include <vector>
-
 namespace Simhash {
-    /**
-     * Compute a similarity hash for the provided string using a rolling hash
-     * function and a provided hash window
-     *
-     * @param tokens - An NULL-terminated array of character pointers to the
-     *                 nul-terminated tokens comprising the text.
-     *
-     * @return hash representative of the content of the text */
-    template <typename Hash=jenkins>
-    class Simhash {
+    class HasherInterface {
+    public:
+        virtual void update(hash_t) = 0;
```

Comment:
It is an interesting question. I do not have definite answer. There is couple considerations:

* Subtype-based polymorphism works better with Python. It is easy to imagine single `simhash` function (or class) that takes moving window, hasher, tokenizer etc. as parameters, and do configuration of the chain. Similar operation with templates is more difficult. (Just in case, I do have a variant of the code with templates.)

* While it is a loop, it is not the inner loop. True consumers are tokenizer and string hash functions (plus, potentially, string -> unicode if needed).

The implemented way was @dlecocq 's suggestion, and if I misunderstood / misstated something I would appreciate him to chime in.